[all hosts] (Storage) Add information about partition local storage and partitionKey (#4399)
OfficeDev/office-js-docs-pr: docs/develop/persisting-add-in-state-and-settings.md

---
title: Persist add-in state and settings
description: Learn how to persist data in Office Web Add-in applications running in the stateless environment of a browser control.
ms.date: 02/08/2024
ms.localizationpriority: medium
---

# Persist add-in state and settings

Office Add-ins are essentially web applications running in the stateless environment of a browser iframe or a webview control. (For brevity hereafter, this article uses "browser control" to mean "browser or webview control".) When in use, your add-in may need to persist data to maintain the continuity of certain operations or features across sessions. For example, your add-in may have custom settings or other values that it needs to save and reload the next time it's initialized, such as a user's preferred view or default location. To do that, you can:

> - [Use the application-specific Office JavaScript APIs for Excel, Word, and Outlook that store data](#application-specific-settings-and-persistence).
> - [Use techniques provided by the underlying browser control](#browser-storage).

Reviewer: ```suggestion
- [Use techniques provided by the underlying browser control](#browser-storage).
- [Use the application-specific Office JavaScript APIs for Excel, Word, and Outlook that store data](#application-specific-settings-and-persistence).
```
Suggestion: Swap to match order of sections in document.

## Application-specific settings and persistence

Excel, Word, and Outlook provide application-specific APIs to save settings and other data. Use these instead of the [Common APIs mentioned later in this article](#common-api-settings-and-persistence) so that your add-in follows consistent patterns and is optimized for the targeted application.

### Settings in Excel and Word

The application-specific JavaScript APIs for Excel and for Word also provide access to the custom settings. Settings are unique to a single Excel file and add-in pairing. For more information, see [Excel.SettingCollection](/javascript/api/excel/excel.settingcollection) and [Word.SettingCollection](/javascript/api/word/word.settingcollection).

The following example shows how to create and access a setting in Excel. The process is functionally equivalent in Word, which uses [Document.settings](/javascript/api/word/word.document#word-word-document-settings-member) instead of `Workbook.settings`.

```js
await Excel.run(async (context) => {

>     let settings = context.workbook.settings;
>     settings.add("NeedsReview", true);
>     let needsReview = settings.getItem("NeedsReview");

Reviewer: ```suggestion
    const settings = context.workbook.settings;
    settings.add("NeedsReview", true);
    const needsReview = settings.getItem("NeedsReview");
```
Nit: Since the variables aren't reassigned new values in the snippet.

    needsReview.load("value");

    await context.sync();
    console.log("Workbook needs review : " + needsReview.value);
});
```

> This section discusses custom XML parts in the context of the Office Common JavaScript API which is supported in Word. The application-specific JavaScript APIs for Excel and for Word also provide access to the custom XML parts. The application-specific APIs and programming patterns are somewhat different from the Common version. For more information, see [Excel.CustomXmlPart](/javascript/api/excel/excel.customxmlpart) and [Word.CustomXmlPart](/javascript/api/word/word.customxmlpart).

Reviewer: Should this be moved to the "How to save custom XML to the document" section?

A [Word.Document](/javascript/api/word/word.document#word-word-document-customxmlparts-member) and [Excel.Workbook](/javascript/api/excel/excel.workbook#excel-excel-workbook-customxmlparts-member) contain a `CustomXmlPartCollection`, which is a list of `CustomXmlParts`. These give access to the XML strings and a corresponding unique ID. By storing these IDs as settings, your add-in can maintain the keys to its XML parts between sessions.

The following samples show how to use custom XML parts with an Excel workbook. The first code block demonstrates how to embed XML data. It stores a list of reviewers, then uses the workbook's settings to save the XML's `id` for future retrieval. The second block shows how to access that XML later. The "ContosoReviewXmlPartId" setting is loaded and passed to the workbook's `customXmlParts`. The XML data is then printed to the console. The process is functionally equivalent in Word, which uses [Document.customXmlParts](/javascript/api/word/word.document#word-word-document-customxmlparts-member) instead of `Workbook.customXmlParts`.

```js
await Excel.run(async (context) => {
    // Add reviewer data to the document as XML

>     let originalXml = "<Reviewers xmlns='http://schemas.contoso.com/review/1.0'><Reviewer>Juan</Reviewer><Reviewer>Hong</Reviewer><Reviewer>Sally</Reviewer></Reviewers>";
>     let customXmlPart = context.workbook.customXmlParts.add(originalXml);
>     customXmlPart.load("id");
>     await context.sync();
> 
>     // Store the XML part's ID in a setting
>     let settings = context.workbook.settings;
>     settings.add("ContosoReviewXmlPartId", customXmlPart.id);
> });
> ```
> 
> ```js
> await Excel.run(async (context) => {
>     // Retrieve the XML part's id from the setting
>     let settings = context.workbook.settings;
>     let xmlPartIDSetting = settings.getItemOrNullObject("ContosoReviewXmlPartId").load("value");
> 
>     await context.sync();
> 
>     if (xmlPartIDSetting.value) {
>         let customXmlPart = context.workbook.customXmlParts.getItem(xmlPartIDSetting.value);
>         let xmlBlob = customXmlPart.getXml();
> 
>         await context.sync();
> 
>         // Add spaces to make it more human-readable in the console.
>         let readableXML = xmlBlob.value.replace(/></g, "> <");

Reviewer: ```suggestion
    const originalXml = "<Reviewers xmlns='http://schemas.contoso.com/review/1.0'><Reviewer>Juan</Reviewer><Reviewer>Hong</Reviewer><Reviewer>Sally</Reviewer></Reviewers>";
    const customXmlPart = context.workbook.customXmlParts.add(originalXml);
    customXmlPart.load("id");
    await context.sync();

    // Store the XML part's ID in a setting
    const settings = context.workbook.settings;
    settings.add("ContosoReviewXmlPartId", customXmlPart.id);
});
```

```js
await Excel.run(async (context) => {
    // Retrieve the XML part's id from the setting
    const settings = context.workbook.settings;
    const xmlPartIDSetting = settings.getItemOrNullObject("ContosoReviewXmlPartId").load("value");

    await context.sync();

    if (xmlPartIDSetting.value) {
        const customXmlPart = context.workbook.customXmlParts.getItem(xmlPartIDSetting.value);
        const xmlBlob = customXmlPart.getXml();

        await context.sync();

        // Add spaces to make it more human-readable in the console.
        const readableXML = xmlBlob.value.replace(/></g, "> <");
```

        console.log(readableXML);
    }
});
```

> [!NOTE]
> `CustomXMLPart.namespaceUri` is only populated if the top-level custom XML element contains the `xmlns` attribute.

### How to save settings in an Outlook add-in

For information about how to save settings in an Outlook add-in, see [Get and set add-in metadata for an Outlook add-in](../outlook/metadata-for-an-outlook-add-in.md) and [Get and set internet headers on a message in an Outlook add-in](../outlook/internet-headers.md).

## Common API settings and persistence

The [Common APIs](understanding-the-javascript-api-for-office.md#api-models) provide objects to save add-in state across sessions, as described in the following table. In all cases, the saved settings values are associated with the [Id](/javascript/api/manifest/id) of the add-in that created them.

> |Object|Add-in type support|Storage location|Office application support|

Reviewer: I think we can remove this table. Aside from Settings and CustomXmlParts, the rest of the APIs in the table are now described in the application-specific section.

If we remove the table, a section describing Excel.CustomProperty and Word.CustomProperty will need to be added to the application-specific section.

Author: Good call. I'll pull that information from the Excel article that talks about custom properties.

- [Use techniques provided by the underlying browser control](#browser-storage).
- [Use the application-specific Office JavaScript APIs for Excel, Word, and Outlook that store data](#application-specific-settings-and-persistence).

If you need to persist state across documents, such as tracking user preferences across any documents they open, you'll need to use a different approach. For example, you could use [SSO](use-sso-to-get-office-signed-in-user-token.md) to obtain the user identity, and then save the user ID and their settings to an online database.

## Browser storage

Persist data across add-in instances with tools from the underlying browser control, such as browser cookies or HTML5 web storage ([localStorage](https://developer.mozilla.org/docs/Web/API/Window/localStorage) or [sessionStorage](https://developer.mozilla.org/docs/Web/API/Window/sessionStorage)).

Some browsers or the user's browser settings may block browser-based storage techniques. You should test for availability as documented in [Using the Web Storage API](https://developer.mozilla.org/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API).

### Storage partitioning

> As a best practice, any private data should be stored in partitioned `localStorage`. [Office.context.partitionKey](/javascript/api/office/office.context#office-office-context-partitionkey-member) provides a local storage partition key only accessible by your add-in. The following example shows how to use the partition key with `localStorage`. Note that the partition key is undefined in environments without partitioning, such as the browser controls for Windows applications.

Reviewer 2: Re: "a local storage partition key only accessible by your add-in"
Strictly, the local Storage (not the key) is only accessible by add-in. A different add-in will get the same key, but when using it in the context of an add-in, only that add-in can access it.
Can we use similar wording like Chrome local storage below like "This means that data stored by storage APIs, such as local storage, are only available to contexts with the same origin and the same top-level site.", 

Not to specific details, can we say something like this here? "This means that data stored by storage APIs, such as local storage, are only available to the same use context."

The `Settings` object is automatically loaded as part of the [Document](/javascript/api/office/office.document) object, and is available when the task pane or content add-in is activated. After the `Document` object is instantiated, you can access the `Settings` object with the [settings](/javascript/api/office/office.document#office-office-document-settings-member) property of the `Document` object. During the lifetime of the session, you can use the `Settings.get`, `Settings.set`, and `Settings.remove` methods to read, write, or remove persisted settings and add-in state from the in-memory copy of the property bag.

Because the set and remove methods operate against only the in-memory copy of the settings property bag, to save new or changed settings back to the document the add-in is associated with, you must call the [Settings.saveAsync](/javascript/api/office/office.settings#office-office-settings-saveasync-member(1)) method.

#### Create or update a setting value

> The following code example shows how to use the [Settings.set](/javascript/api/office/office.settings#office-office-settings-set-member(1)) method to create a setting called `'themeColor'` with a value `'green'`. The first parameter of the set method is the case-sensitive  _name_ (Id) of the setting to set or create. The second parameter is the _value_ of the setting.

Reviewer: ```suggestion
The following code example shows how to use the [Settings.set](/javascript/api/office/office.settings#office-office-settings-set-member(1)) method to create a setting called `'themeColor'` with a value `'green'`. The first parameter of the set method is the case-sensitive  *name* (Id) of the setting to set or create. The second parameter is the *value* of the setting.
```

The setting with the specified name is created if it doesn't already exist, or its value is updated if it does exist. Use the `Settings.saveAsync` method to persist the new or updated settings to the document.

#### Get the value of a setting

> The following example shows how use the [Settings.get](/javascript/api/office/office.settings#office-office-settings-get-member(1)) method to get the value of a setting called "themeColor". The only parameter of the `get` method is the case-sensitive _name_ of the setting.

Reviewer: ```suggestion
The following example shows how use the [Settings.get](/javascript/api/office/office.settings#office-office-settings-get-member(1)) method to get the value of a setting called "themeColor". The only parameter of the `get` method is the case-sensitive *name* of the setting.
```

> The `get` method returns the value that was previously saved for the setting _name_ that was passed in. If the setting doesn't exist, the method returns **null**.

Reviewer: ```suggestion
The `get` method returns the value that was previously saved for the setting *name* that was passed in. If the setting doesn't exist, the method returns **null**.
```

#### Remove a setting

> The following example shows how to use the [Settings.remove](/javascript/api/office/office.settings#office-office-settings-remove-member(1)) method to remove a setting with the name "themeColor". The only parameter of the `remove` method is the case-sensitive _name_ of the setting.

Reviewer: ```suggestion
The following example shows how to use the [Settings.remove](/javascript/api/office/office.settings#office-office-settings-remove-member(1)) method to remove a setting with the name "themeColor". The only parameter of the `remove` method is the case-sensitive *name* of the setting.
```

Nothing will happen if the setting doesn't exist. Use the `Settings.saveAsync` method to persist removal of the setting from the document.

#### Save your settings

> To save any additions, changes, or deletions your add-in made to the in-memory copy of the settings property bag during the current session, you must call the [Settings.saveAsync](/javascript/api/office/office.settings#office-office-settings-saveasync-member(1)) method to store them in the document. The only parameter of the `saveAsync` method is _callback_, which is a callback function with a single parameter.

Reviewer: ```suggestion
To save any additions, changes, or deletions your add-in made to the in-memory copy of the settings property bag during the current session, you must call the [Settings.saveAsync](/javascript/api/office/office.settings#office-office-settings-saveasync-member(1)) method to store them in the document. The only parameter of the `saveAsync` method is *callback*, which is a callback function with a single parameter.
```

> The anonymous function passed into the `saveAsync` method as the _callback_ parameter is executed when the operation is completed. The _asyncResult_ parameter of the callback provides access to an `AsyncResult` object that contains the status of the operation. In the example, the function checks the `AsyncResult.status` property to see if the save operation succeeded or failed, and then displays the result in the add-in's page.

Reviewer: ```suggestion
The anonymous function passed into the `saveAsync` method as the *callback* parameter is executed when the operation is completed. The *asyncResult* parameter of the callback provides access to an `AsyncResult` object that contains the status of the operation. In the example, the function checks the `AsyncResult.status` property to see if the save operation succeeded or failed, and then displays the result in the add-in's page.
```
